README
This setup is for working without the Arduino IDE.

Instead, it's command line based, using the make utility.
All the files needed for development under Windows are included.
The build still uses Arduino 0022's core libraries.
The toolchain is the one used by Arduino 0022.
The core libraries and toolchain are located in the arduino-0022 subdirectory.

A batch file, make.bat, is supplied to call up the included version of make.

Before trying to build for the first time, and every time that you download a new Makefile from the repository, or add a new source file to Makefile, run
  make depend

To build open_evse.hex:
  make
To load the firmware via a USBasp programmer:
  make fuses
To load the firmware via a USBasp programmer:
  make flash
To clean up intermediate files from the build process:
  make clean

Put any added libraries in the libraries subdirectory.

I've commented out the basic configuration #defines from open_evse.h. You
can control them via the CDEFS lines in Makefile without having to edit open_evse.h


Firmware for Open EVSE: Open Source Hardware J1772 Electric Vehicle Supply Equipment
See http://code.google.com/p/open-evse/ for project details.

NOTES:
- Working versions of the required libraries are included in the libraries/ folder
------------

Open EVSE is free software; you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation; either version 3, or (at your option)
any later version.

Open EVSE is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with Open EVSE; see the file COPYING.  If not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330,
Boston, MA 02111-1307, USA.

* Open EVSE is distributed in the hope that it will be useful,
* but WITHOUT ANY WARRANTY; without even the implied warranty of
* MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.	

CHANGELOG
Change Log

vD3.6.3 SCL 20150403
- fix compile errors in Arduino 1.5.8+ due to deprecation of prog_char
	

20150327 SCL
- take VERSION out of Makefile.. set it in open_evse.h again
- fix bug in Makefile gcc was being passed wrong flags
- save a few bytes on startup message

vD3.6.2 SCL 20150326
- add delay(200) to beginning of setup() to fix random LCD black out at boot
	Thanks to CraigK!
- code cleanup on KWH_RECORDING .. change eeprom_read_byte() to eeprom_read_dword()
	
vD3.6.1 SCL 20150311
- merge CraigK patch-4 temp monitoring fixes
	
vD3.6.0 SCL 20150301
- add CraigK TEMPERATURE_MONITORING
	
vD3.5.2 SCL 20150226
- minor fixes for millis() overflow handling
- change code to allow RED_LED_PIN and GREEN_LED_PIN to be commented out w/o causing compile errors
	
SCL 20150220
- merge NSayer pull request OpenEVSE II hardware now only performs ground test during charging
	
vD3.5.1 SCL 20150218
- added FT_GFI_LOCKOUT/FT_GFI_RETRY for functional testing
- per UL Fred Reyes, stuck relay always results in hard fault (lockout)
- GFI_RETRY_COUNT to 255 (infinite) instead of 15

vD3.5.0 SCL 20150216
- added UL_COMPLIANT: UL wants EVSE to hard fault (cease operation until manual intervention)
  1) if POST fails
  2) if fault happens immediately after charge starts
- consolidated some redundant fault strings
- if AutoSvcLevel couldn't be detected due to car being plugged in during POST,
  and L1 was set because of this, upon first charge session, we will detect the
  actual svc level, and switch to L2 if detected

vD3.4.3 SCL 20150213
- don't do GFI self test in doPost() if other fault already found
- when in menus, lock out all other functions until exit 
- do stuck relay chk in POST even if EV is plugged in and AutoSvcLevel()
- for safety, set to L1 if AutoSvcLevel and car is plugged in during POST,
  because we can't auto detect level.

vD3.4.2 SCL 20150210
- added SLEEP_STATUS_PIN
	
vD3.4.1 SCL 20150205
- forgot to set svcState to SR when stuck relay test fails in doPost() and
	not AutoSvcLevel()
	
vD3.4.0 SCL 20150123
- added writing of safety check trip counters to EEPROM
- added RAPI $GF
- setting GFI_RETRY_COUNT to 255 will make it retry indefinitely
- changed GFI_RETRY_COUNT from 3 to 15, GFI_TIMEOUT from 15min to 1min
  -> now, tests will retry 15 times at one minute intervals
     instead of 3 retries at 15 min intervals
	
vD3.3.7 SCL 20150122
- merged NSayer pull requests #19,#20
  - OpenEVSE II changes + added relay check menu item
	
vD3.3.6 SCL 20150120
- enable WATCHDOG by default
- minor refactoring of watchdog code
- perform relay stuck check in doPost() even if not auto svc level
	
vD3.3.5 SCL 20150113
- merge NSAYER pull request for OpenEVSE II code
	
vD3.3.4 SCL 20150102
- switch from EEPROM.h to avr/eeprom.h functions
	
vD3.3.3 SCL 20150102
- clear charging current in chargingOff() - thanks to CapnBry
	
vD3.3.2 SCL 20141205
- added very long press RAPI code for WiFi AP mode reset ($WF)
	
vD3.3.1 SCL 20141124
- fix some random crashes caused by string overflow on LCD display
  one such crash was when I2CLCD enabled and RTC enabled w/o RTC hardware
	
vD3.3.0 SCL 20141120
- promote version to 3.3.x since features were added
	
vD3.2.9 SCL 20141120
- make ammeter calibration mode volatile
	
vD3.2.8 SCL 20141119
- restructure and fix ammeter bugs
- don't update charging amps display unless ammeter value is updated
- don't show charging amps if m_CurrentScaleFactor == 0
- replace moving average w/ sliding window average to save RAM
- added ammeter calibration RAPI function $S2
- replace redundant calls to millis() with variable in J1772EVSEController::Update()
	
vD3.2.7 SCL 20141111
- fixed ammeter no display bug introduced in 3.2.6

v3.2.6 SCL 20141111
- restructured ammeter code
- add RAPI SF,SA,GA,GG commands

v3.2.5 SCL 20141111
- don't display charging current <1A
	
v3.2.4 SCL 20141110
- minor fixes to RAPI
- added CraigK's current averaging AMMETER_AVERAGING
	
v3.2.3 SCL 20140829
- added support for PCF8574-based I2C LCD backpacks
	
v3.2.2 SCL 20140829
- setting RGB/mono backlight fixes
	
v3.2.1 SCL 20140826
- fix POST problem when LCD not attached
- remove superfluous m_menuidx added in last update
	
v3.2.0 SCL 20140825
- revamped SetupMenu to greatly simplify adding/removing items
- added LCD backlight type setting to menus (C. Howell)
	
v3.1.0 SCL 20140822
- added CHARGING_PINAC
- promote from alpha to production status
	
v3.1.A9 SCL 20140418
- fixed bug in saving GFI self test setting
	
v3.1.A8 SCL 20140417
- added S.Low's (scanchain) CheckTime() fixes
	
v3.1.A7 SCL 20140417
- bumped version to 3.x.x to avoid confusion w/ main branch code
	
v2.1.A7 SCL 20140417
- added nsayer's GFI_SELFTEST + menu tweaks
	
v2.0.B8 20131118
- merge v2.0.B7 from open_evse google group + some minor tweaks
	
v2.0.B3 20130714
- fix LCD display error
- merged changes from Scott Rubin
	
v2.0.B1 20130119
- added RTC + delay timer code from Goldserve
	
v1.0.9 SCL 20130101
- don't hang when I2CLCD/RGBLCD enabled, and no LCD installed. requires
  LiquidTWI2 v1.2.0+
	
v1.0.8 SCL 20121025
- change button reading from analogRead() to digitalRead(), enable internal pullup
	
v1.0.7 SCL 20121023
- fix bug in Max Current Menu.. couldn't display highest selectable value
  on init
	
v1.0.6 SCL
- switch LCD library to LiquidTWI2 to save memory, speed up LCD
  writing, and decrease latency
	
v1.0.5 SCL
- move g_psEnabled/g_psEnabled to #ifdef SERIALCLI to fix compile
  error when LCD and BTN_MENU disabled
	
v1.0.4 SCL
- BTN_ENABLE_TOGGLE reverts code from v1.0.3
- refine duty cycle calc in SetPWM() (thanks to Lee Howard)
	
v1.0.3 SCL
- short press no longer does anything when menus not active
  instead of emergency start/stop
- button menus only available in connected and charging states
	
v1.0.0
First official release

rapi_proc.cpp
// -*- C++ -*-
/*
 * Open EVSE Firmware
 *
 * Copyright (c) 2013-2014 Sam C. Lin <lincomatic@gmail.com>
 *
 * This file is part of Open EVSE.

 * Open EVSE is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.

 * Open EVSE is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.

 * You should have received a copy of the GNU General Public License
 * along with Open EVSE; see the file COPYING.  If not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "open_evse.h"

#ifdef RAPI
extern J1772EVSEController g_EvseController;
extern OnboardDisplay g_OBD;

const char RAPI_VER[] PROGMEM = RAPIVER;


// convert 2-digit hex string to uint8
uint8 htou8(const char *s)
{
  uint8 u = 0;
  for (int i=0;i < 2;i++) {
    if (i == 1) u <<= 4;
    char c = s[i];
    if ((c >= '0') && (c <= '9')) {
      u += c - '0';
    }
    else if ((c >= 'A') && (c <= 'F')) {
      u += c - 'A' + 10;
    }
  }
  return u;
}

// convert decimal string to uint8
uint8 dtou8(const char *s)
{
  uint8 u = 0;
  while (*s) {
    u *= 10;
    u += *(s++) - '0';
  }
  return u;
}

EvseRapiProcessor::EvseRapiProcessor()
{
  echo = 0;
  reset();
}

//extern HardwareSerial Serial;
int EvseRapiProcessor::doCmd(int8 sendstatetrans)
{
  int rc = 1;

  // chk for state transition and send async notification
  if (sendstatetrans && g_EvseController.StateTransition()) {
    sendEvseState();
  }

  int bcnt = available();
  if (bcnt) {
    for (int i=0;i < bcnt;i++) {
      char c = read();
      if (echo) write(c);

      if (c == ESRAPI_SOC) {
	buffer[0] = ESRAPI_SOC;
	bufCnt = 1;
      }
      else if (buffer[0] == ESRAPI_SOC) {
	if (bufCnt < ESRAPI_BUFLEN) {
	  if (c == ESRAPI_EOC) {
	    buffer[bufCnt++] = 0;
	    if (!tokenize()) {
	      rc = processCmd();
	    }
	    else {
	      reset();
	      response(0);
	    }
	  }
	  else {
	    buffer[bufCnt++] = c;
	  }
	}
	else { // too many chars
	  reset();
	}
      }
    }
  }

  return rc;
}

void EvseRapiProcessor::sendEvseState()
{
  extern char g_sTmp[64];
  sprintf(g_sTmp,"%cST %02x%c",ESRAPI_SOC,g_EvseController.GetState(),ESRAPI_EOC);
  write(g_sTmp);
}

void EvseRapiProcessor::setWifiMode(uint8_t mode)
{
  extern char g_sTmp[64];
  sprintf(g_sTmp,"%cWF %02x%c",ESRAPI_SOC,(int)mode,ESRAPI_EOC);
  write(g_sTmp);
}

int EvseRapiProcessor::tokenize()
{
  tokens[0] = &buffer[1];
  char *s = &buffer[2];
  tokenCnt = 1;
  uint8 chkSum = ESRAPI_SOC + buffer[1];
  uint8 ichkSum = 0;
  while (*s) {
    if (*s == ' ') {
      chkSum += *s;
      *s = '\0';
      tokens[tokenCnt++] = ++s;
    }
    else if (*s == '*') {
      *(s++) = '\0';
      ichkSum = htou8(s);
	  break;
    }
    else {
      chkSum += *(s++);
    }
  }

  return (chkSum == ichkSum) ? 0 : 1;
}

int EvseRapiProcessor::processCmd()
{
  int rc = -1;
  unsigned u1,u2,u3;
  int i,i2;
  uint8 x,y;

  // we use bufCnt as a flag in response() to signify data to write
  bufCnt = 0;

  char *s = tokens[0];
  switch(*(s++)) { 
  case 'F': // function
    switch(*s) {
#ifdef LCD16X2
    case 'B': // LCD backlight
      g_OBD.LcdSetBacklightColor(dtou8(tokens[1]));
      rc = 0;
      break;
#endif // LCD16X2      
    case 'D': // disable EVSE
      g_EvseController.Disable();
      rc = 0;
      break;
    case 'E': // enable EVSE
      g_EvseController.Enable();
      rc = 0;
      break;
#ifdef LCD16X2
    case 'P': // print to LCD
      {
	u1 = dtou8(tokens[1]); // x
	u2 = dtou8(tokens[2]); // y
	// now restore the spaces that were replaced w/ nulls by tokenizing
	for (i=4;i < tokenCnt;i++) {
	  *(tokens[i]-1) = ' ';
	}
	g_OBD.LcdPrint(u1,u2,tokens[3]);
      }
      rc = 0;
      break;
#endif // LCD16X2      
    case 'R': // reset EVSE
      extern void WatchDogResetEvse();
      WatchDogResetEvse();
      rc = 0;
      break;
    case 'S': // sleep
      g_EvseController.Sleep();
      rc = 0;
      break;
    default:
      ; // do nothing
    }
    break;

  case 'S': // set parameter
    switch(*s) {
#ifdef LCD16X2
    case '0': // set LCD type
      if (tokenCnt == 2) {
#ifdef RGBLCD
	rc = g_EvseController.SetBacklightType((*tokens[1] == '0') ? BKL_TYPE_MONO : BKL_TYPE_RGB);
#endif // RGBLCD
      }
      break;
#endif // LCD16X2      
#ifdef RTC      
    case '1': // set RTC
      if (tokenCnt == 7) {
	extern void SetRTC(uint8 y,uint8 m,uint8 d,uint8 h,uint8 mn,uint8 s);
	SetRTC(dtou8(tokens[1]),dtou8(tokens[2]),dtou8(tokens[3]),
	       dtou8(tokens[4]),dtou8(tokens[5]),dtou8(tokens[6]));
	rc = 0;
      }
      break;
#endif // RTC      
#ifdef AMMETER
    case '2': // ammeter calibration mode
      if (tokenCnt == 2) {
	g_EvseController.EnableAmmeterCal((*tokens[1] == '1') ? 1 : 0);
	rc = 0;
      }
      break;
    case 'A':
      if (tokenCnt == 3) {
	sscanf(tokens[1],"%d",&i);
	g_EvseController.SetCurrentScaleFactor(i);
	sscanf(tokens[2],"%d",&i);
	g_EvseController.SetAmmeterCurrentOffset(i);
	rc = 0;
      }
      break;
#endif // AMMETER
    case 'C': // current capacity
      if (tokenCnt == 2) {
	rc = g_EvseController.SetCurrentCapacity(dtou8(tokens[1]),1);
      }
      break;
    case 'D': // diode check
      if (tokenCnt == 2) {
	g_EvseController.EnableDiodeCheck((*tokens[1] == '0') ? 0 : 1);
	rc = 0;
      }
      break;
    case 'E': // echo
      if (tokenCnt == 2) {
	echo = ((*tokens[1] == '0') ? 0 : 1);
	rc = 0;
      }
      break;
#ifdef GFI_SELFTEST
    case 'F': // GFI self test
      if (tokenCnt == 2) {
	g_EvseController.EnableGfiSelfTest(*tokens[1] == '0' ? 0 : 1);
	rc = 0;
      }
      break;
#endif // GFI_SELFTEST
#ifdef ADVPWR
    case 'G': // ground check
      if (tokenCnt == 2) {
	g_EvseController.EnableGndChk(*tokens[1] == '0' ? 0 : 1);
	rc = 0;
      }
      break;
#endif // ADVPWR
    case 'L': // service level
      if (tokenCnt == 2) {
	switch(*tokens[1]) {
	case '1':
	case '2':
	  g_EvseController.SetSvcLevel(*tokens[1] - '0',1);
#ifdef ADVPWR
	  g_EvseController.EnableAutoSvcLevel(0);
#endif
	  rc = 0;
	  break;
#ifdef ADVPWR
	case 'A':
	  g_EvseController.EnableAutoSvcLevel(1);
	  rc = 0;
	  break;
#endif // ADVPWR
	default:
	  ;
	}
      }
      break;
#ifdef ADVPWR      
    case 'R': // stuck relay check
      if (tokenCnt == 2) {
	g_EvseController.EnableStuckRelayChk(*tokens[1] == '0' ? 0 : 1);
	rc = 0;
      }
      break;
#endif // ADVPWR      
#ifdef GFI_SELFTEST
    case 'S': // GFI self-test
      if (tokenCnt == 2) {
	g_EvseController.EnableGfiSelfTest(*tokens[1] == '0' ? 0 : 1);
	rc = 0;
      }
      break;
#endif // GFI_SELFTEST   
#ifdef DELAYTIMER     
    case 'T': // timer
      if (tokenCnt == 5) {
	extern DelayTimer g_DelayTimer;
	if ((*tokens[1] == '0') && (*tokens[2] == '0') && (*tokens[3] == '0') && (*tokens[4] == '0')) {
	  g_DelayTimer.Disable();
	}
	else {
	  g_DelayTimer.SetStartTimer(dtou8(tokens[1]),dtou8(tokens[2]));
	  g_DelayTimer.SetStopTimer(dtou8(tokens[3]),dtou8(tokens[4]));
	  g_DelayTimer.Enable();
	}
	rc = 0;
      }
      break;
#endif // DELAYTIMER      
    case 'V': // vent required
      if (tokenCnt == 2) {
	g_EvseController.EnableVentReq(*tokens[1] == '0' ? 0 : 1);
	rc = 0;
      }
      break;
    }
    break;

  case 'G': // get parameter
    switch(*s) {
#ifdef AMMETER
    case 'A':
      i = g_EvseController.GetCurrentScaleFactor();
      i2 = g_EvseController.GetAmmeterCurrentOffset();
      sprintf(buffer,"%d %d",i,i2);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
#endif // AMMETER
    case 'C': // get current capacity range
      sprintf(buffer,"%d %d",MIN_CURRENT_CAPACITY,(g_EvseController.GetCurSvcLevel() == 2) ? MAX_CURRENT_CAPACITY_L2 : MAX_CURRENT_CAPACITY_L1);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
    case 'E': // get settings
      u1 = g_EvseController.GetCurrentCapacity();
      u2 = g_EvseController.GetFlags();
      sprintf(buffer,"%d %04x",u1,u2);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
    case 'F': // get fault counters
      u1 = eeprom_read_byte((uint8_t*)EOFS_GFI_TRIP_CNT);
      u2 = eeprom_read_byte((uint8_t*)EOFS_NOGND_TRIP_CNT);
      u3 = eeprom_read_byte((uint8_t*)EOFS_STUCK_RELAY_TRIP_CNT);
      sprintf(buffer,"%x %x %x",u1,u2,u3);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
#ifdef AMMETER
    case 'G':
      u1 = g_EvseController.GetChargingCurrent();
      sprintf(buffer,"%u",u1);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
#endif // AMMETER
    case 'S': // get state
      sprintf(buffer,"%d %ld",g_EvseController.GetState(),g_EvseController.GetElapsedChargeTime());
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
#ifdef RTC
    case 'T': // get time
      extern void GetRTC(char *buf);
      GetRTC(buffer);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
#endif // RTC
    case 'V': // get version
      extern void GetVerStr(char *buf);
      GetVerStr(buffer);
      strcat(buffer," ");
      strcat_P(buffer,RAPI_VER);
      bufCnt = 1; // flag response text output
      rc = 0;
      break;
    default:
      ; //do nothing
    }

  default:
    ; // do nothing
  }

  response((rc == 0) ? 1 : 0);

  reset();

  return rc;
}

void EvseRapiProcessor::response(uint8 ok)
{
  write(ESRAPI_SOC);
  write(ok ? "OK " : "NK ");

  if (bufCnt) {
    write(buffer);
  }
  write(ESRAPI_EOC);
}


EvseRapiProcessor g_ERP;
#endif // RAPI
